### PR TITLE
Implement async `postproc_pagebreaks`

### PR DESCRIPTION
**Pull request by lafrenierejm**
This is intended to address one of the todos from #146:
> Fixing / Converting the postprocessors to async. Specifically… `postproc_pagebreaks` in https://github.com/phiresky/ripgrep-all/blob/master/src/adapters/postproc.rs .

This should definitely receive thorough review before acceptance. I am _not_ an experienced Rust developer, and wasn't able to make use of the unit tests in `postproc.rs` because (AFAICT) they are broken pending fixes in other functions.

### CODE REVIEW

File: src/adapters/postproc.rs [hunk 1]
```diff
@@ -145,30 +143,35 @@ pub fn postproc_prefix(line_prefix: &str, inp: impl AsyncRead + Send) -> impl As
...
+        yield Ok(Bytes::copy_from_slice(line_prefix.as_bytes()));
+        for await chunk in inp_stream {
+            match chunk {
+                Err(e) => yield Err(e),
+                Ok(chunk) => {
+                    let chunk_iter = chunk.split(|byte| byte == &form_feed);
+                    for sub_chunk in chunk_iter {
+                        if sub_chunk.contains(&b'\n') {
+                            yield Ok(Bytes::copy_from_slice(&regex.replace_all(&sub_chunk, line_prefix.as_bytes())));
+                            page_count += 1;
```
Comment (phiresky) [inline src/adapters/postproc.rs:166]: i think you have a logic error here. the page count should increase for each chunk here not just if the chunk contains \n.
or rather, it should increase on each chunk except the first one i guess

Comment (lafrenierejm) [reply src/adapters/postproc.rs:164]: Right you are. I tried again in f19e173585c454355b9221059fed4dc1a9115792, where `page_count` is incremented when and only when a form feed character is encountered.

Comment (phiresky) [reply src/adapters/postproc.rs:164]: uh i think that's still not right :D 

The stream would look like this (`\f` is form feed)

`Hello\nWorld\fFoo Bar\n\fTest`

splitting would result in:

`Hello\nWorld`,   `Foo Bar\n`, `Test`

The result should be

Page 1: Hello\nWorld
Page 2: Foo Bar\n
Page 3: Test

You don't just want to increase the page if the chunk is *just* form_feed (I think)

Comment (phiresky) [reply src/adapters/postproc.rs:164]: (updated, my example was wrong

File: src/adapters/postproc.rs [hunk 2]
```diff
@@ -145,38 +143,57 @@ pub fn postproc_prefix(line_prefix: &str, inp: impl AsyncRead + Send) -> impl As
...
- * where N starts at one and is incremented for each ASCII Form Feed character in the input stream.
- * (That's the format output by pdftotext)
- */
-pub fn postproc_pagebreaks(line_prefix: &str, inp: impl AsyncRead) -> impl AsyncRead {
-    let line_prefix = line_prefix.to_string(); // clone since
-    let mut page_count = 1;
+/// Adds the prefix "Page N:" to each line,
+/// where N starts at one and is incremented for each ASCII Form Feed character in the input stream.
+/// ASCII form feeds are the page delimiters output by `pdftotext`.
+pub fn postproc_pagebreaks(input: impl AsyncRead + Send) -> impl AsyncRead + Send {
```
Comment (lafrenierejm) [inline src/adapters/postproc.rs:149]: I couldn't get this working with the `line_prefix: &str` parameter. Compiler was complaining about an implicit lifetime on that parameter that I couldn't figure out.

File: src/adapters/postproc.rs [hunk 3]
```diff
@@ -145,38 +143,57 @@ pub fn postproc_prefix(line_prefix: &str, inp: impl AsyncRead + Send) -> impl As
...
+                        page_prefix = format!("\nPage {}:", page_count);
+                    }
+                }
             }
-            _ => b"[[imposs]]".to_vec(),
-        }),
-    }*/
+        }
+    };
+    Box::pin(StreamReader::new(output_stream))
```
Comment (lafrenierejm) [inline src/adapters/postproc.rs:173]: I'm not sure whether the use of `Box::pin` is strictly necessary here. It was just the first thing I tried that satisfied the compiler, so I haven't dug further into understanding pinning.

File: src/adapters/postproc.rs (hunk 2 above)
Comment (lafrenierejm) [reply src/adapters/postproc.rs]: Added in https://github.com/lafrenierejm/ripgrep-all/commit/4cac831a8f6a8f308ef940fdcc79cc92175009a5.

File: src/adapters/postproc.rs (hunk 3 above)
Comment (phiresky) [reply src/adapters/postproc.rs:178]: that's ok i think, i don't have a deep understanding of pinning / unpinning either

**Review by phiresky — APPROVED** [review]
thanks, looks pretty good now! 
i hope you didn't spend too much time on the line_prefix thing because that could potentially also be done by chianing postproc_pagebreaks and postproc_prefix.

the next step here will be to figure out how/when this adapter is actually invoked. i think it should probably happen with the normal file-type detection, making `.txtwithpagebreaks` a file extension that's detected by this adapter. for that CustomAdapterConfig needs a field for the output file extension, set to `output_filename: "$input_filename.txtwithpagebreaks`

i'll try to implement that

**Comment by lafrenierejm** [comment]
> i hope you didn't spend too much time on the line_prefix thing because that could potentially also be done by chianing postproc_pagebreaks and postproc_prefix.

No, I didn't spend much time on the line prefix param. I figured chaining would be an option if needed, but wanted to avoid forcing that unless the larger architectural concerns warranted it.

> the next step here will be to figure out how/when this adapter is actually invoked. i think it should probably happen with the normal file-type detection, making `.txtwithpagebreaks` a file extension that's detected by this adapter. for that CustomAdapterConfig needs a field for the output file extension, set to `output_filename: "$input_filename.txtwithpagebreaks`
> 
> 
> 
> i'll try to implement that

Sounds good!

**Pull request merged**